src/_c_internal_utils.c

```diff
@@ -1,11 +1,61 @@
 #define PY_SSIZE_T_CLEAN
 #include <Python.h>
+#ifdef __linux__
+#include <dlfcn.h>
+#endif
 #ifdef _WIN32
 #include <Objbase.h>
 #include <Shobjidl.h>
 #include <Windows.h>
 #endif
 
+static PyObject* mpl_display_is_invalid(PyObject* module)
+{
+#ifdef __linux__
```

Comment:
I think it would still make sense to check that environment variables DISPLAY, or WAYLAND_DISPLAY exist before hand. First, to avoid libwayland's behavior of defaulting to connect to 'wayland-0', even if no environment variables are set; second, because dlopen is very slow -- one one computer, calling `dlopen` followed by `dlclose` on `libX11.so` and `libwayland-client.so` costs between 1.4 msec (marginal cost), 2 msec (initial cost, hot cache), and 6 msec (uncached, initial cost). Importing matplotlib already costs me 240msec every script load. (Aside: Why not check env variable WAYLAND_SOCKET? Because the socket named by WAYLAND_SOCKET cannot be reused.)

Reply:
Can you check the speed of getenv()?  (I genuinely do not know whether it is much faster than dlopen().)

Reply:
getenv is very fast, since the environment variables are provided to a program at the start, just like the command line arguments. My benchmark of getenv uses around 0.0002 msec to call both getenv("WAYLAND_DISPLAY") and getenv("DISPLAY").

Reply:
ok, then.